File: www/site/cgi-bin/weight_convert.php
#!/usr/bin/php
<?php
// Gravity factors relative to Earth
$gravity_factors = [
    "mercury" => 0.38,
    "venus"   => 0.91,
    "earth"   => 1.0,
    "mars"    => 0.38,
    "jupiter" => 2.34,
    "saturn"  => 1.06,
    "uranus"  => 0.92,
    "neptune" => 1.14,
    "pluto"   => 0.06,
    "moon"    => 0.16
];

try {
    // PHP automatically parses both GET and POST into $_REQUEST
    $weight_val = $_REQUEST['weight'] ?? null;
    $planet_val = $_REQUEST['planet'] ?? null;

    if (!$weight_val || !$planet_val) {
        throw new Exception("Missing parameters");
    }

    $planet_name = strtolower($planet_val);
    if (!isset($gravity_factors[$planet_name])) {
        throw new Exception("Invalid planet");
    }

    // Calculations
    $earth_weight = (float)$weight_val;
    $factor = $gravity_factors[$planet_name];
    $final_weight = round($earth_weight * $factor, 3);

    // Load Template
    $template_path = __DIR__ . "/result.html";
    if (!file_exists($template_path)) {
        throw new Exception("Template not found");
    }
    $html = file_get_contents($template_path);

    // Replace Placeholders
    $html = str_replace("{{FINAL_WEIGHT}}", number_format($final_weight, 3, '.', ''), $html);
    $html = str_replace("{{PLANET_NAME}}", strtoupper($planet_name), $html);
    $html = str_replace("{{EARTH_WEIGHT}}", number_format($earth_weight, 3, '.', ''), $html);

    // Output for your C++ Parser
    // PHP's 'header()' function won't work correctly in a raw CGI script 
    // being parsed by a custom C++ server, so we print manually:
    echo "Status: 200 OK\r\n";
    echo "Content-Type: text/html\r\n";
    echo "Content-Length: " . strlen($html) . "\r\n";
    echo "\r\n"; // Header/Body separator
    echo $html;

} catch (Exception $e) {
    echo "Status: 500 Internal Server Error\r\n";
    echo "Content-Type: text/plain\r\n";
    echo "\r\n";
    echo "Error: " . $e->getMessage();
}
?>

File: www/site/cgi-bin/weight_convert.py
#!/usr/bin/env python3
import os
import sys
import urllib.parse

# Gravity factors relative to Earth
GRAVITY_FACTORS = {
    "mercury": 0.38, "venus": 0.91, "earth": 1.0,
    "mars": 0.38, "jupiter": 2.34, "saturn": 1.06,
    "uranus": 0.92, "neptune": 1.14, "pluto": 0.06, "moon": 0.16
}

def get_input_data():
    """Manual replacement for cgi.FieldStorage()"""
    method = os.environ.get("REQUEST_METHOD", "GET")
    
    if method == "POST":
        content_length = int(os.environ.get("CONTENT_LENGTH", 0))
        query_string = sys.stdin.read(content_length)
    else:
        query_string = os.environ.get("QUERY_STRING", "")
    
    # Parses 'weight=100&planet=mars' into {'weight': ['100'], 'planet': ['mars']}
    return urllib.parse.parse_qs(query_string)

def main():
    try:
        data = get_input_data()
        
        # 1. Handle Inputs
        weight_val = data.get("weight", [None])[0]
        planet_val = data.get("planet", [None])[0]

        if not weight_val or not planet_val:
            # If inputs are missing, we send a partial response that 
            # your C++ 'if(res.body.empty()...)' check will catch
            return 

        planet_name = planet_val.lower()
        earth_weight = float(weight_val)
        final_weight = earth_weight * GRAVITY_FACTORS.get(planet_name, 1.0)

        # 2. Prepare the Body
        script_dir = os.path.dirname(os.path.abspath(__file__))
        with open(os.path.join(script_dir, "result.html"), "r") as f:
            template = f.read()

        body = template.replace("{{FINAL_WEIGHT}}", f"{final_weight:.3f}")
        body = body.replace("{{PLANET_NAME}}", planet_name.upper())
        body = body.replace("{{EARTH_WEIGHT}}", f"{earth_weight:.3f}")

        # 3. Output EXACTLY what parseCgiOutput expects
        # Note: We provide Status and Content-Type so your C++ struct fills correctly
        sys.stdout.write("Status: 200 OK\r\n")
        sys.stdout.write("Content-Type: text/html\r\n")
        sys.stdout.write(f"Content-Length: {len(body)}\r\n")
        sys.stdout.write("\r\n") # The critical separator
        sys.stdout.write(body)
        sys.stdout.flush()

    except Exception as e:
        # This will trigger your ERROR_LOG("Malformed CGI output...")
        pass 

if __name__ == "__main__":
    main()